src/session_pool/session.js

```diff
@@ -0,0 +1,182 @@
+     * @param options.userData {Object} - Object where custom user data can be stored. For example custom headers.
+     * @param options.maxErrorScore {number} - Maximum number of failed session usage.
+     * If the `errorScore` reaches the `maxErrorScore` session is marked as block and it is thrown away.
+     * @param options.errorScoreDecrement {number} - It is used for healing the session.
+     * For example: if your session fails two times, but it is successful on the third attempt it's errorScore is decremented by this number.
+     * @param options.createdAt {Date} - Date of creation.
+     * @param options.expiredAt {Date} - Date of expiration.
+     * @param options.usageCount {Number} - Indicates how many times the session has been used.
+     * @param options.errorCount {Number} - Indicates how many times the session failed.
+     * @param options.maxSessionUsageCount {Number} - Session should be used only a limited amount of times.
+     * This number indicates how many times the session is going to be used, before it is thrown away.
+     * @param options.sessionPool {EventEmitter} - SessionPool instance. Session will emit the `sessionRetired` event on this instance.
+     */
+    constructor(options = {}) {
+        const {
+            id = `session-${cryptoRandomObjectId(10)}`,
+            cookies = [],
+            maxAgeSecs = 3000,
+            userData = {},
+            maxErrorScore = 3,
+            errorScoreDecrement = 0.5,
+            createdAt = new Date(),
+            usageCount = 0,
+            errorScore = 0,
+            maxSessionUsageCount = 50,
+            sessionPool,
+        } = options;
+
+        let { expiresAt } = options;
+
+        // Validation
+        checkParamOrThrow(id, 'options.id', 'String');
+        checkParamOrThrow(maxAgeSecs, 'options.maxAgeSecs', 'Number');
+        checkParamOrThrow(userData, 'options.userData', 'Object');
+        checkParamOrThrow(maxErrorScore, 'options.maxErrorScore', 'Number');
+        checkParamOrThrow(expiresAt, 'options.expiresAt', 'Maybe Date');
+        checkParamOrThrow(createdAt, 'options.createdAt', 'Date');
+        checkParamOrThrow(usageCount, 'options.usageCount', 'Number');
+        checkParamOrThrow(errorScore, 'options.errorScore', 'Number');
+        checkParamOrThrow(maxSessionUsageCount, 'options.maxSessionUsageCount', 'Number');
+        checkParamOrThrow(sessionPool, 'options.sessionPool', 'Object');
+
+        // sessionPool must be at least instance of EvenEmitter.
+        // That way we can allow custom implementation of SessionPool in the future (It should not be needed).
+        if (!(sessionPool instanceof EventEmitter)) {
+            throw new Error('Session: sessionPool must be instance of SessionPool');
+        }
+
+        if (!expiresAt) {
+            expiresAt = new Date((new Date()).getTime() + (maxAgeSecs * 1000));
```

Comment:
I'd simplify this to `new Date(Date.now() + maxAgeSecs * 1000)` and add it as default value instead of having this `if` clause.